django/db/models/lookups.py

```diff
@@ -190,6 +198,53 @@ class FieldGetDbPrepValueIterableMixin(FieldGetDbPrepValueMixin):
     """
     get_db_prep_lookup_value_is_iterable = True
 
+    def get_prep_lookup(self):
+        prepared_values = []
+        if hasattr(self.rhs, '_prepare'):
+            # A subquery is like an iterable, but we shouldn't independently
+            # prepare its items
+            return self.rhs._prepare(self.lhs.output_field)
+        for rhs_value in self.rhs:
+            if hasattr(rhs_value, 'resolve_expression'):
+                # A combinable expression will be handled by the database, but
+                # can co-exist alongside real values
+                pass
+            elif self.prepare_rhs and hasattr(self.lhs.output_field, 'get_prep_value'):
+                rhs_value = self.lhs.output_field.get_prep_value(rhs_value)
+            prepared_values.append(rhs_value)
+        return prepared_values
+
+    def process_rhs(self, compiler, connection):
+        if self.rhs_is_direct_value():
+            # rhs should be an iterable of values, we use batch_process_rhs
+            # to prepare/transform those values
+            return self.batch_process_rhs(compiler, connection)
+        else:
+            return super(FieldGetDbPrepValueIterableMixin, self).process_rhs(compiler, connection)
+
+    def resolve_combinable_parameter(self, compiler, connection, sql, param):
+        params = [param]
+        if hasattr(param, 'resolve_expression'):
+            param = param.resolve_expression(compiler.query)
+        if hasattr(param, 'as_sql'):
+            sql, params = param.as_sql(compiler, connection)
+        return sql, params
+
+    def batch_process_rhs(self, compiler, connection, rhs=None):
+        pre_processed = super(FieldGetDbPrepValueIterableMixin, self).batch_process_rhs(compiler, connection, rhs)
+        # The params list we have may contain expressions, which themselves compile to a sql/param pair. We zip them
+        # together to get sql and param pairs that refer to the same argument, and attempt to replace them with the
+        # result of compiling the param step.
+        params = []
+        for sql, param in zip(*pre_processed):
+            params.append(self.resolve_combinable_parameter(compiler, connection, sql, param))
+        sql, params = zip(*params)
+        params = sum(params, [])
+        # The TO_DELETE singleton represents parameters that have been integrated directly into sql, so they are
+        # replaced to avoid a mismatch of number of format strings and parameters.
+        params = [item for item in params if item is not TO_DELETE]
+        return sql, params
```

Comment:
You can reduce iterations of params by unpacking in a single pass:

```
        params = []
        for sql, param in zip(*pre_processed):
            params.append(self.resolve_combinable_parameter(compiler, connection, sql, param))
        sql, params = zip(*params)
        # The TO_DELETE singleton represents parameters that have been integrated directly into sql, so they are
        # replaced to avoid a mismatch of number of format strings and parameters.
        params = [item for subparams in params for item in subparams if item is not TO_DELETE]
        return sql, params
```

While I was looking at this though I noticed that `TO_DELETE` isn't actually used anywhere. Is this left over from something you've changed since? If it's not needed then can you simplify the return of `resolve_combinable_parameter` such that you don't need to rezip into sql, params?